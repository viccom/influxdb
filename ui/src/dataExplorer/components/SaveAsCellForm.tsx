import React, {PureComponent, ChangeEvent} from 'react'
import {connect} from 'react-redux'

// utils
import _ from 'lodash'

// components
import {
  Grid,
  Form,
  Input,
  Button,
  ComponentColor,
  ButtonType,
  Columns,
  ComponentStatus,
  InputType,
} from 'src/clockface'
import {ErrorHandling} from 'src/shared/decorators/errors'
import DashboardsDropdown from 'src/dataExplorer/components/DashboardsDropdown'

// constants
import {cellAddFailed, cellAdded} from 'src/shared/copy/notifications'

// actions
import {getDashboardsAsync, createCellWithView} from 'src/dashboards/actions/v2'
import {createDashboard} from 'src/dashboards/apis/v2'
import {notify} from 'src/shared/actions/notifications'

// types
import {AppState, Dashboard, View, Organization} from 'src/types/v2'
import {
  DashboardTemplate,
  DEFAULT_DASHBOARD_NAME,
  DEFAULT_CELL_NAME,
} from 'src/dashboards/constants'

interface State {
  targetDashboardIDs: string[]
  cellName: string
  isNameDashVisible: boolean
  newDashboardName: string
}

interface StateProps {
  dashboards: Dashboard[]
  view: View
  orgs: Organization[]
}

interface DispatchProps {
  handleGetDashboards: typeof getDashboardsAsync
  onCreateCellWithView: typeof createCellWithView
  notify: typeof notify
}

interface OwnProps {
  dismiss: () => void
}

type Props = StateProps & DispatchProps & OwnProps

@ErrorHandling
class SaveAsCellForm extends PureComponent<Props, State> {
  public state: State = {
    targetDashboardIDs: [],
    cellName: '',
    isNameDashVisible: false,
    newDashboardName: '',
  }

  public async componentDidMount() {
    const {handleGetDashboards} = this.props
    await handleGetDashboards()
  }

  public render() {
    const {dismiss, dashboards} = this.props
    const {cellName, isNameDashVisible, targetDashboardIDs} = this.state
    return (
      <Form onSubmit={this.handleSubmit}>
        <Grid>
          <Grid.Row>
            <Grid.Column widthXS={Columns.Twelve}>
              <Form.Element label="Target Dashboard(s)">
                <DashboardsDropdown
                  onSelect={this.handleSelectDashboardID}
                  selectedIDs={targetDashboardIDs}
                  dashboards={dashboards}
                />
              </Form.Element>
            </Grid.Column>
            {isNameDashVisible && this.nameDashboard}
            <Grid.Column widthXS={Columns.Twelve}>
              <Form.Element label="Cell Name">
                <Input
                  type={InputType.Text}
                  placeholder="Add optional cell name"
                  name="cellName"
                  value={cellName}
                  onChange={this.handleChangeCellName}
                />
              </Form.Element>
            </Grid.Column>
            <Grid.Column widthXS={Columns.Twelve}>
              <Form.Footer>
                <Button
                  text="Cancel"
                  onClick={dismiss}
                  titleText="Cancel save"
                  type={ButtonType.Button}
                />
                <Button
                  text={'Save as Dashboard Cell'}
                  color={ComponentColor.Success}
                  type={ButtonType.Submit}
                  onClick={this.handleSubmit}
                  status={
                    this.isFormValid
                      ? ComponentStatus.Default
                      : ComponentStatus.Disabled
                  }
                />
              </Form.Footer>
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </Form>
    )
  }

  private get nameDashboard(): JSX.Element {
    const {newDashboardName} = this.state
    return (
      <Grid.Column widthXS={Columns.Twelve}>
        <Form.Element label="New Dashboard Name">
          <Input
            type={InputType.Text}
            placeholder="Add dashboard name"
            name="dashboardName"
            value={newDashboardName}
            onChange={this.handleChangeDashboardName}
          />
        </Form.Element>
      </Grid.Column>
    )
  }

  private get isFormValid(): boolean {
    const {targetDashboardIDs} = this.state
    return !_.isEmpty(targetDashboardIDs)
  }

  private handleSubmit = async () => {
    const {onCreateCellWithView, dashboards, view, dismiss, notify} = this.props
    const {targetDashboardIDs} = this.state

    const cellName = this.state.cellName || DEFAULT_CELL_NAME
    const newDashboardName =
      this.state.newDashboardName || DEFAULT_DASHBOARD_NAME

    const viewWithProps: View = {...view, name: cellName}

    try {
      targetDashboardIDs.forEach(dashID => {
        let targetDashboardName = ''
        try {
          if (dashID === DashboardTemplate.id) {
            targetDashboardName = newDashboardName
            this.handleCreateDashboardWithView(newDashboardName, viewWithProps)
          } else {
            const selectedDashboard = dashboards.find(d => d.id === dashID)
            targetDashboardName = selectedDashboard.name
            onCreateCellWithView(selectedDashboard, viewWithProps)
          }
          notify(cellAdded(cellName, targetDashboardName))
        } catch {
          notify(cellAddFailed(cellName, targetDashboardName))
        }
      })
    } finally {
      this.resetForm()
      dismiss()
    }
  }

  private handleCreateDashboardWithView = async (
    dashboardName: string,
    view: View
  ): Promise<void> => {
    const {onCreateCellWithView, orgs} = this.props
    try {
      const newDashboard = {
        orgID: orgs[0].id,
        name: dashboardName || DEFAULT_DASHBOARD_NAME,
        cells: [],
      }
      const dashboard = await createDashboard(newDashboard)
      onCreateCellWithView(dashboard, view)
    } catch (error) {
      console.error(error)
    }
  }

  private resetForm() {
    this.setState({
      targetDashboardIDs: [],
      cellName: '',
      isNameDashVisible: false,
      newDashboardName: '',
    })
  }

  private handleSelectDashboardID = (
    selectedIDs: string[],
    value: Dashboard
  ) => {
    if (value.id === DashboardTemplate.id) {
      this.setState({
        isNameDashVisible: selectedIDs.includes(DashboardTemplate.id),
      })
    }
    this.setState({targetDashboardIDs: selectedIDs})
  }

  private handleChangeDashboardName = (e: ChangeEvent<HTMLInputElement>) => {
    this.setState({newDashboardName: e.target.value})
  }

  private handleChangeCellName = (e: ChangeEvent<HTMLInputElement>) => {
    this.setState({cellName: e.target.value})
  }
}

const mstp = (state: AppState): StateProps => {
  const {
    orgs,
    dashboards,
    timeMachines: {
      timeMachines: {de},
    },
  } = state
  const {view} = de
  return {dashboards, view, orgs}
}

const mdtp: DispatchProps = {
  handleGetDashboards: getDashboardsAsync,
  onCreateCellWithView: createCellWithView,
  notify,
}

export default connect<StateProps, DispatchProps>(
  mstp,
  mdtp
)(SaveAsCellForm)
