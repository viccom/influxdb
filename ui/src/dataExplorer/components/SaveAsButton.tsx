// Libraries
import React, {PureComponent} from 'react'

// Components
import SaveAsCellForm from 'src/dataExplorer/components/SaveAsCellForm'
import SaveAsTaskForm from 'src/dataExplorer/components/SaveAsTaskForm'
import {
  IconFont,
  Button,
  Radio,
  OverlayTechnology,
  OverlayBody,
  OverlayHeading,
  OverlayContainer,
  ComponentColor,
} from 'src/clockface'

// Styles
import 'src/dataExplorer/components/SaveAsButton.scss'

enum SaveAsOption {
  Dashboard = 'dashboard',
  Task = 'task',
}

interface Props {}

interface State {
  isOverlayVisible: boolean
  saveAsOption: SaveAsOption
}

class SaveAsButton extends PureComponent<Props, State> {
  public state: State = {
    isOverlayVisible: false,
    saveAsOption: SaveAsOption.Dashboard,
  }

  public render() {
    const {isOverlayVisible, saveAsOption} = this.state

    return (
      <>
        <Button
          icon={IconFont.Export}
          text="Save As"
          onClick={this.handleShowOverlay}
          color={ComponentColor.Primary}
          titleText="Save your query as a Dashboard Cell or a Task"
        />
        <OverlayTechnology visible={isOverlayVisible}>
          <OverlayContainer maxWidth={600}>
            <OverlayHeading
              title="Save As"
              onDismiss={this.handleHideOverlay}
            />
            <OverlayBody>
              <div className="save-as--options">
                <Radio>
                  <Radio.Button
                    active={saveAsOption === SaveAsOption.Dashboard}
                    value={SaveAsOption.Dashboard}
                    onClick={this.handleSetSaveAsOption}
                    data-test="cell-radio-button"
                  >
                    Dashboard Cell
                  </Radio.Button>
                  <Radio.Button
                    active={saveAsOption === SaveAsOption.Task}
                    value={SaveAsOption.Task}
                    onClick={this.handleSetSaveAsOption}
                    data-test="task-radio-button"
                  >
                    Task
                  </Radio.Button>
                </Radio>
              </div>
              {this.saveAsForm}
            </OverlayBody>
          </OverlayContainer>
        </OverlayTechnology>
      </>
    )
  }

  private get saveAsForm(): JSX.Element {
    const {saveAsOption} = this.state

    if (saveAsOption === SaveAsOption.Dashboard) {
      return <SaveAsCellForm dismiss={this.handleHideOverlay} />
    } else if (saveAsOption === SaveAsOption.Task) {
      return <SaveAsTaskForm dismiss={this.handleHideOverlay} />
    }
  }

  private handleShowOverlay = () => {
    this.setState({isOverlayVisible: true})
  }

  private handleHideOverlay = () => {
    this.setState({isOverlayVisible: false})
  }

  private handleSetSaveAsOption = (saveAsOption: SaveAsOption) => {
    this.setState({saveAsOption})
  }
}

export default SaveAsButton
