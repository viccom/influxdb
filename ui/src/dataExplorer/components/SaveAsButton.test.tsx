// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import SaveAsButton from 'src/dataExplorer/components/SaveAsButton'
import SaveAsCellForm from 'src/dataExplorer/components/SaveAsCellForm'

const setup = () => {
  const wrapper = shallow(<SaveAsButton />)

  return {wrapper}
}

describe('SaveAsButton', () => {
  const {wrapper} = setup()
  describe('rendering', () => {
    it('renders', () => {
      expect(wrapper.exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('save as cell form', () => {
    it('defaults to save as cell form', () => {
      const saveAsCellForm = wrapper.find(SaveAsCellForm)

      expect(saveAsCellForm.exists()).toBe(true)
    })
  })
})
