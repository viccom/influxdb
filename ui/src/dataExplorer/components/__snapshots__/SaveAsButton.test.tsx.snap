// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveAsButton rendering renders 1`] = `
<Fragment>
  <Button
    active={false}
    color="primary"
    icon="export"
    onClick={[Function]}
    shape="none"
    size="sm"
    status="default"
    text="Save As"
    titleText="Save your query as a Dashboard Cell or a Task"
    type="button"
  />
  <OverlayTechnology
    visible={false}
  >
    <OverlayContainer
      maxWidth={600}
    >
      <OverlayHeading
        onDismiss={[Function]}
        title="Save As"
      />
      <OverlayBody>
        <div
          className="save-as--options"
        >
          <Radio
            color="primary"
            shape="none"
            size="sm"
          >
            <RadioButton
              active={true}
              data-test="cell-radio-button"
              disabled={false}
              disabledTitleText="This option is disabled"
              onClick={[Function]}
              value="dashboard"
            >
              Dashboard Cell
            </RadioButton>
            <RadioButton
              active={false}
              data-test="task-radio-button"
              disabled={false}
              disabledTitleText="This option is disabled"
              onClick={[Function]}
              value="task"
            >
              Task
            </RadioButton>
          </Radio>
        </div>
        <Connect(SaveAsCellForm)
          dismiss={[Function]}
        />
      </OverlayBody>
    </OverlayContainer>
  </OverlayTechnology>
</Fragment>
`;
